### DISCUSSION TITLE
InnerProductSpace working example

### DISCUSSION THREAD

**Discussion by MohammadForouhesh** [non/spire#1308] [Q&A] [answered]
We plan to use this wonderful library in one of our service, yet I couldn't find any proper tutorial on how to run a simple example. our use case is as follows: 
Lets say we have a case class Person, how can I calculate the dot product between two Person?
```
case class Person(age: Int, height: Int) 
val p = Person(20, 180)
val c = Person(15, 165)
p.dot(c)
```
If you provide me with some sample code, it would be highly appreciated.

**Comment by armanbilge** [comment, marked as answer]
Are you familiar with implicits/givens/typeclasses? Are you using Scala 2 or Scala 3? You want to define an instance of [`InnerProductSpace`](https://github.com/typelevel/spire/blob/1006dbf6b5e3122bfc234f6e05a3b6b532ab0f70/core/src/main/scala/spire/algebra/InnerProductSpace.scala#L19) and place inside the companion `object Person`.

Here's an example for Scala 3.

```scala
object Person {
  given InnerProductSpace[Person, Int] with {
    def dot(x: Person, y: Person): Int = ...
  }
}
```

Then to get the dot syntax you would do something like `import spire.syntax.all._`

**Reply by MohammadForouhesh** [reply]
thanks for your timely reply and your sample code. also I wonder whether there is any benchmark on the performance of this library against other possible solutions?
